src/runtime-rs/crates/resource/src/network/endpoint/physical_endpoint.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::path::Path;
+
+use anyhow::{Context, Result};
+use async_trait::async_trait;
+use hypervisor::{device, Hypervisor};
+
+use super::Endpoint;
+use crate::network::utils::{self, link};
+
+pub const SYS_PCI_DEVICES_PATH: &str = "/sys/bus/pci/devices";
+
+#[derive(Debug)]
+pub struct VendorDevice {
+    vendor_id: String,
+    device_id: String,
+}
+
+impl VendorDevice {
+    pub fn new(vendor_id: &str, device_id: &str) -> Self {
+        Self {
+            vendor_id: vendor_id.to_string(),
+            device_id: device_id.to_string(),
+        }
+    }
+
+    pub fn vendor_device_id(&self) -> String {
+        let vendor_device_id = format!("{} {}", &self.vendor_id, &self.device_id);
```

Comment:
Can vendor_id or device_id hold spaces? If not, `format!` will return a String, so you need the String::from and trim.

Reply:
Do you mean 'don't' need ?

Reply:
Thanks for your review, change to `format!("{}_{}", &self.vendor_id, &self.device_id)` maybe more clearly.